### PR TITLE
Chapter 5 Code updates

### PR DESCRIPTION
**Pull request by MarkMichaelis**

### CODE REVIEW

**Comment by BenjaminMichaelis** [comment]
Looks like you need some rebasing maybe

**Comment by Keboo** [comment]
I followed up on teams and pushed a "betterChapter5" branch as an example.

File: src/Chapter05.Tests/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter05.Tests;
+
+public class Constants
+{
+    public const string Inigo = "Inigo";
+    public const string Montoya = "Montoya";
```
Comment (Keboo) [inline src/Chapter05.Tests/Constants.cs:6]: Not saying you should do this but there is a fun trick here.
```suggestion
    public const string Inigo = nameof(Inigo);
    public const string Montoya = nameof(Montoya);
```

File: src/Chapter05.Tests/Listing05.15.Tests.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        });
+    }
+
+    [TestMethod]
+    public void Main_IntelliTectIndexHtmlArgs_DownloadFile()
+    {
+        try
+        { 
+            string[] args = { "http://IntelliTect.com", "Index.html" };
+            string expected = $"Downloaded 'Index.html' from '{ args[0]}'.";
```
Comment (Keboo) [inline src/Chapter05.Tests/Listing05.15.Tests.cs]: Is it a problem that this hits a 404?
[https://IntelliTect.com/Index.html](https://IntelliTect.com/Index.html)

File: src/Chapter05.Tests/Listing05.15.Tests.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+            int? result = null;
+            IntelliTect.TestTools.Console.ConsoleAssert.Expect(
+                expected, () => result = Program.Main(args));
+        }
+        catch (AggregateException exception) when (exception.InnerException!.GetType() == typeof(System.Net.Http.HttpRequestException))
+        {
+            Assert.Inconclusive("Unable to download the file.  Check your Internet connection.");
```
Comment (Keboo) [inline src/Chapter05.Tests/Listing05.15.Tests.cs]:
```suggestion
            Assert.Inconclusive("Unable to download the file.  Check your internet connection.");
```

File: src/Chapter05.Tests/Listing05.15.Tests.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        try
+        { 
+            string[] args = { "http://IntelliTect.com", "Index.html" };
+            string expected = $"Downloaded 'Index.html' from '{ args[0]}'.";
+
+            int? result = null;
+            IntelliTect.TestTools.Console.ConsoleAssert.Expect(
+                expected, () => result = Program.Main(args));
+        }
+        catch (AggregateException exception) when (exception.InnerException!.GetType() == typeof(System.Net.Http.HttpRequestException))
```
Comment (Keboo) [inline src/Chapter05.Tests/Listing05.15.Tests.cs]:
```suggestion
        catch (AggregateException exception) when (exception.InnerException is System.Net.Http.HttpRequestException)
```

File: src/Chapter05.Tests/Listing05.15.Tests.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    [TestMethod]
+    public void Main_GoodArgs_DownloadFile()
+    {
+        string[] args = { "http://google.com", Path.Combine(Directory.GetCurrentDirectory(), "destination.txt") };
+
+        try
+        {
+            Assert.AreEqual(0, Program.Main(args));
+        }
+        catch (AggregateException exception) when (exception.InnerException!.GetType() == typeof(System.Net.Http.HttpRequestException))
```
Comment (Keboo) [inline src/Chapter05.Tests/Listing05.15.Tests.cs]:
```suggestion
        catch (AggregateException exception) when (exception.InnerException is System.Net.Http.HttpRequestException)
```

File: src/Chapter05.Tests/Listing05.15.Tests.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            Assert.Inconclusive("Unable to download the file.  Check your Internet connection.");
```
Comment (Keboo) [inline src/Chapter05.Tests/Listing05.15.Tests.cs:54]:
```suggestion
            Assert.Inconclusive("Unable to download the file.  Check your internet connection.");
```

File: src/Chapter05/Listing05.16.PassingVariablesByValue.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+using System;
+
+#region INCLUDE
+public class Program
+{
+    public static void Main()
+    {
+        // ...
+        string fullName;
+        string driveLetter = "C:";
```
Comment (Keboo) [inline src/Chapter05/Listing05.16.PassingVariablesByValue.cs:12]: This is a very Windows-centric example. Would it be sufficient to simply show combining directories and drop the volume specifier. It also opens up the question of the colon (`Path.VolumeSeparatorChar`) after "C" which is only true on Windows and Mac, and a `/` on *nix systems.

File: src/Chapter05.Tests/Listing05.20.Tests.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        string view = Path.Combine(Directory.GetCurrentDirectory(), "bin", "config", "index.html");
+        view += Environment.NewLine;
+        view += Path.Combine(Environment.SystemDirectory, "Temp", "index.html");
+        view += Environment.NewLine;
+        view += $"C:{Path.DirectorySeparatorChar}{Path.Combine("Data", "HomeDir", "index.html")}";
+
+//            string view =
+//Directory.GetCurrentDirectory() + @"\bin\config\index.html
+//" + Directory.GetParent(Directory.GetCurrentDirectory()).FullName + @"\Temp\index.html
+//C:\Data\HomeDir\index.html";
```
Comment (Keboo) [inline src/Chapter05.Tests/Listing05.20.Tests.cs:21]:
```suggestion
```

File: src/Chapter05/Listing05.20.PassingAVariableParameterList.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        #endregion HIGHLIGHT
+        Console.WriteLine(fullName);
+
+        // ...
+
+        #region HIGHLIGHT
+        // Call Combine() with an array
+        fullName = Combine(
+            new string[] {
+                $"C:{Path.DirectorySeparatorChar}", "Data",
```
Comment (Keboo) [inline src/Chapter05/Listing05.20.PassingAVariableParameterList.cs]: Can we drop the `C:` here and use something like the previous two examples so it is not locked into Windows?

File: src/Chapter05.Tests/Listing05.21.Tests.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+[TestClass]
+public class ProgramTests
+{
+    [TestMethod]
+    public void Main_WritePath()
+    {
+        string currentDirectory = Environment.CurrentDirectory;
+
+        int expected=0;
```
Comment (Keboo) [inline src/Chapter05.Tests/Listing05.21.Tests.cs]: nit spacing.
```suggestion
        int expected = 0;
```

File: src/Chapter05.Tests/Listing05.27.Tests.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+            {
+                ExceptionHandling.Main();
+            });
+        }
+        catch (FormatException)
+        {
+            flag = true;
+        }
+
+        Assert.IsTrue(flag);
```
Comment (Keboo) [inline src/Chapter05.Tests/Listing05.27.Tests.cs:30]: Logged an issue to make this pattern cleaner.
https://github.com/IntelliTect/TestTools.Console/issues/20

File: src/Chapter05/Chapter05.csproj [hunk 12]
```diff
@@ -8,6 +8,12 @@
     <ExpectedWarnings>$(ExpectedWarnings),CS1058</ExpectedWarnings>
   </PropertyGroup>
   <ItemGroup>
+    <None Remove="Listing05.07.ImplicitUsingsGeneratedGlocalUsingDeclaratives.cs" />
+    <None Remove="Listing05.09.UsingElementInProjectFile.xml" />
+  </ItemGroup>
+  <ItemGroup>
+    <Using Include="System.Net" />
+    <!--<Using Include="System.Timers.Timer" Alias="Timer" />-->
```
Comment (Keboo) [inline src/Chapter05/Chapter05.csproj]:
```suggestion
```

File: src/Chapter05/Listing05.29.ThrowingAnException.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        try
+        {
+            Console.WriteLine("Begin executing");
+            Console.WriteLine("Throw exception");
+            #region HIGHLIGHT
+            throw new Exception("Arbitrary exception");
+            // Catch 1
+            #endregion HIGHLIGHT
+            Console.WriteLine("End executing");
+            }
```
Comment (Keboo) [inline src/Chapter05/Listing05.29.ThrowingAnException.cs]: nit: whitespace
```suggestion
        }
```

**Review by Keboo — APPROVED** [review]

**Review by Keboo — APPROVED** [review]

File: src/Chapter05/Listing05.07.ImplicitUsingsGeneratedGlocalUsingDeclaratives.cs [hunk 14]
```diff
@@ -0,0 +1,10 @@
+#region INCLUDE
+// <auto-generated/>
+global using global::System;
```
Comment (BenjaminMichaelis) [inline src/Chapter05/Listing05.07.ImplicitUsingsGeneratedGlocalUsingDeclaratives.cs:3]: File name is mispelled

File: src/Chapter05/Listing05.16.PassingVariablesByValue.cs [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+using System;
+
+#region INCLUDE
+public class Program
+{
+    public static void Main()
+    {
+        // ...
+        string fullName;
+        string homeFolder = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
```
Comment (BenjaminMichaelis) [inline src/Chapter05/Listing05.16.PassingVariablesByValue.cs]:
```suggestion
        // Environment.SpecialFolder.UserProfile on different OS
        // WIN: C:\Users\intellitect
        // LIN: /home/intellitect
        // OSX: /Users/intellitect
        string homeFolder = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
```
could be fun to explain the result for each os

File: src/Chapter05/Listing05.28.GeneralCatchBlocks.cs [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        try
+        {
+            age = int.Parse(ageText);
+            Console.WriteLine(
+                $"Hi { firstName }!  You are { age * 12 } months old.");
+        }
+        catch(FormatException exception)
+        {
+            Console.WriteLine(
+                $"The age entered ,{ageText}, is not valid.");
```
Comment (BenjaminMichaelis) [inline src/Chapter05/Listing05.28.GeneralCatchBlocks.cs:36]:
```suggestion
                $"The age entered, {ageText}, is not valid.");
```

File: src/Chapter05/Listing05.28.GeneralCatchBlocks.cs [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        Console.Write("Enter your age: ");
+        // Assume not null for clarity
+        ageText = Console.ReadLine()!;
+
+        #endregion EXCLUDE
+        try
+        {
+            age = int.Parse(ageText);
+            Console.WriteLine(
+                $"Hi { firstName }!  You are { age * 12 } months old.");
```
Comment (BenjaminMichaelis) [inline src/Chapter05/Listing05.28.GeneralCatchBlocks.cs]:
```suggestion
                $"Hi { firstName }! You are { age * 12 } months old.");
```

File: src/Chapter05/Listing05.31.TryCatch.cs [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+        if (int.TryParse(ageText, out int age))
+        {
+            Console.WriteLine(
+                $"Hi { firstName }! " +
+                $"You are { age * 12 } months old.");
+        }
+        else
+        {
+            Console.WriteLine(
+                $"The age entered ,{ ageText }, is not valid.");
```
Comment (BenjaminMichaelis) [inline src/Chapter05/Listing05.31.TryCatch.cs]:
```suggestion
                $"The age entered, { ageText }, is not valid.");
```

**Review by BenjaminMichaelis — APPROVED** [review]

**Pull request merged**